dropshot/src/router.rs

```diff
@@ -501,7 +539,41 @@ pub fn path_to_segments(path: &str) -> Vec<&str> {
      * much here, but it does create more work, so we'll just split it
      * ourselves.
      */
-    path.split('/').filter(|segment| !segment.is_empty()).collect::<Vec<_>>()
+    let mut ret = Vec::new();
+    for segment in path.0.split('/') {
+        if !segment.is_empty() {
+            ret.push(percent_decode_str(segment).decode_utf8()?.to_string());
+        }
+    }
+    Ok(ret)
```

Comment:
Completely optional nit, if we want to have one less `mut`:

```suggestion
   Ok(
     path.0.split('/')
       .filter(|segment| !segment.is_empty())
       .map(|segment| {
         Ok(percent_decode_str(segment).decode_utf8()?.to_string())
       })
       .collect()
   );
```

Reply:
I had something a bit like that, but I struggled to elegantly deal with the embedded `Result`. I really like this trick of returning `Ok()` to force the closure to have a `Result` type -- that's cool. Also cool that `Result` implements `FromIterator`. I think this work if I remove the surrounding `Ok(...)`. Thanks!!